daphne/src/testing.rs

```diff
@@ -341,10 +348,66 @@ impl DapLeader<BearerToken> for MockAggregator {
 
     async fn get_reports(
         &self,
-        _task_id: &Id,
-        _selector: &Self::ReportSelector,
+        task_id: &Id,
+        selector: &Self::ReportSelector,
     ) -> Result<Vec<Report>, DapError> {
-        unreachable!("not implemented");
+        let task_config = self
+            .get_task_config_for(task_id)
+            .ok_or_else(|| DapError::fatal("no task found"))?;
+
+        // Calculate batch interval.
+        let batch_interval = if let Some(ref interval) = selector.batch_info {
+            if !interval.is_valid_for(task_config) {
+                return Err(DapError::fatal("invalid batch interval"));
+            }
+            interval.clone()
+        } else {
+            task_config.current_batch_window(Utc::now().timestamp() as u64)
+        };
+
+        // Do some checks on batch interval.
+        if batch_interval.start > Utc::now().timestamp() as u64 {
+            return Err(DapError::fatal(
+                "invalid batch interval: start time cannot be in the future",
+            ));
+        }
+
+        // Lock report_store.
+        let agg_rate = selector.agg_rate as usize;
+        let mut reports = Vec::with_capacity(agg_rate);
+        let mut report_store_mutex_guard = self
+            .report_store
+            .lock()
+            .map_err(|e| DapError::Fatal(e.to_string()))?;
+        let report_store = report_store_mutex_guard.deref_mut();
+
+        // Fetch reports.
+        let mut window = batch_interval.start;
+        // TODO(nakatsuka-y) Should the batch window be inclusive?
+        while window < batch_interval.end() {
+            let num_reports_remaining = agg_rate - reports.len();
+            let bucket_info = BucketInfo {
+                task_id: task_id.clone(),
+                window,
+            };
+
+            // If there are reports in this bucket, append into reports.
+            // If not, move on to the next bucket.
+            if let Some(ref mut store) = report_store.get_mut(&bucket_info) {
+                reports.append(&mut store.process_get_pending(num_reports_remaining as u64));
+                match reports.len().cmp(&agg_rate) {
+                    Ordering::Less => (),
+                    Ordering::Greater => return Err(DapError::fatal(
+                        "number of reports received from report_store exceeds the number requested",
+                    )),
+                    Ordering::Equal => break,
+                }
```

Comment:
Hmm, this `match` statement strikes me as somewhat awkward. In particular, since `store` is in memory (and not implemented by some external service, as is the case for the DO code), why do we need the `store_proecess_getpending()` callback at all? Why not access `store`'s memory directly? This would let us simplify this code a great deal.

Reply:
I've replaced `store.process_get_pending` with directly accessing `pending`.
I think the code looks cleaner now.